File: src/java/main/other/RichJsonEnvironment.java
package other;

import core.RichJsonConstants;

import java.util.HashMap;
import java.util.Map;

/**
 * Verwaltet die globalen Umgebungsvariablen (Macros) für RichJson.
 */
public class RichJsonEnvironment {

    public static final Map<String, Object> env = new HashMap<>();

    /**
     * Fügt mehrere Konstanten zur Umgebung hinzu.
     * @param envs Eine Map mit Namen und Werten.
     */
    public static void addEnvironmentVariables(Map<String, Object> envs) {
        if (envs == null) return;

        for (Map.Entry<String, Object> entry : envs.entrySet()) {
            addEnvironmentVariable(entry.getKey(), entry.getValue());
        }
    }

    /**
     * Fügt eine einzelne Konstante zur Umgebung hinzu.
     * @param name Der Name der Variable.
     * @param value Der Wert der Variable.
     */
    public static void addEnvironmentVariable(String name, Object value) {
        if (RichJsonConstants.KEY_COMMAND_MEMBER.equals(name)) {
            return;
        }

        if (env.containsKey(name)) {
            throw new RuntimeException("RichJson has the macro '" + name + "' already defined");
        }

        env.put(name, value);
    }
}

File: src/c++/main/other/RichJsonEnvironment.cpp
#include "RichJsonEnvironment.hpp"
#include "../core/RichJsonConstants.hpp"
#include <stdexcept>

namespace RichJson {

void RichJsonEnvironment::addEnvironmentVariable(const std::string& name, const json& value) {
    if (name == RichJsonConstants::KEY_COMMAND_MEMBER) return;

    if (hasEnv(name) || hasFunction(name)) {
        throw std::runtime_error("RichJson has the env variable '" + name + "' already defined");
    }

    env()[name] = value;
}

void RichJsonEnvironment::addEnvironmentVariables(const json& envs) {
    if (envs.is_null()) return;

    for (auto it = envs.begin(); it != envs.end(); ++it) {
        addEnvironmentVariable(it.key(), it.value());
    }
}

void RichJsonEnvironment::registerFunction(const std::string& name, std::function<json()> fn) {
    if (hasEnv(name) || hasFunction(name)) {
        throw std::runtime_error("RichJson has the env variable '" + name + "' already defined");
    }

    functions()[name] = std::move(fn);
}

}
